processing/src/main/java/org/apache/druid/query/GenericQueryMetricsFactory.java

```diff
@@ -52,11 +50,7 @@
   /**
    * Creates a {@link QueryMetrics} which doesn't have predefined QueryMetrics subclass. This method is used
    * by the router to build a {@link QueryMetrics} for SQL queries. It is needed since at router, there is no native
-   * query linked to a SQL query. By default, it returns null to maintain backward compatibility.
+   * query linked to a SQL query.
    */
-  @Nullable
-  default QueryMetrics<Query<?>> makeMetrics()
-  {
-    return null;
-  }
+  QueryMetrics<Query<?>> makeMetrics();
```

Comment:
do we not need a default implementation?

Reply:
I removed the default implementation since it is ok as per the `PublicApi` documentation to add method to interfaces in a major release. Further, the existing method's semantics are kept same.

Reply:
Just because it is "okay" to add a thing, forcing anybody that happens to implement an interface to implement a new method adds friction to moving forward.  It's "okay" to provide for *some* way to make changes and move things forward, not to say that it's okay to force people to make simple random changes to keep up with versions.  Always make it as simple as possible for someone to bring their extension forward a version, if it's relatively easy to make it so that nobody has to change anything and they get good behavior, then that should be done.

Now, if the default implementation doesn't provide good behavior and instead causes bad things to happen, then people should be forced to implement it.  In this case, it seems like a default is good?

Reply:
I'm not sure if the default would be a good thing to have here. having a default method in the interface which returns `null` would mean that all the developers who have written custom `GenericQueryMetricsFactory` would need to implement this method anyways to get router metrics. Also this information to extend the factory would need communication.
Further since the method return would be a nullable, all users of the method would need to handle null explicitly in the code from now on.
The good thing that happens with default implementation is that developers who don't care about router metrics for SQL queries don't need to make any code changes to their custom implementation of metrics factory.

Reply:
Then the answer isn't "I don't have a default implementation because it's okay to add one" it's "I don't have one because there isn't a good implementation for them and anybody who implements this interface really does need to think about what the correct implementation is".  In which case, that's the reason to add a new method, so great :).

Anyway, sorry to be pedantic, but for anything that impacts compatibility, it's important to show the work that we've thought about the plight of the developer who is updating their cluster.

Reply:
Yes, I totally agree with your sentiment regarding SPI compat being more of a judgement call rather than a technicality. I started out with a default implementation and then removed it due to the above rationale. But I missed adding the full explanation in the previous comment.
After this discussion I was also thinking if we could change `makeMetrics(Query)` to `makeMetrics(@Nullable Query)`  and then make the `makeMetrics` default implementation as `makeMetrics(null)`. But again, with that I think that might impact the semantics of `makeMetrics(Query)` to expect a non-null `Query`.
I'll again think if any other way is possible to avoid incompatbility and update if I find one.

Reply:
I believe this is resolved.  It is the way it is in this PR because that's the way it needs to be.